app/hooks/use-money-input.ts

```diff
@@ -0,0 +1,216 @@
+import { useEffect, useState } from 'react';
+import type { NumpadButton } from '~/components/numpad';
+import type { Ticker } from '~/lib/exchange-rate';
+import type { Currency } from '~/lib/money';
+import { Money, getLocaleDecimalSeparator } from '~/lib/money';
+import { getUnit } from '~/utils';
+import { useExchangeRates } from './use-exchange-rate';
+
+type State = {
+  input: {
+    /** Raw input value */
+    value: string;
+    /** Currency of the input */
+    currency: Currency;
+  };
+  converted: {
+    /** Raw converted value. Undefined if rates are not loaded yet */
+    value: string | undefined;
+    /** Currency of the converted value */
+    currency: Currency;
+  };
+};
+
+type Props = {
+  /** Initial input value as raw string */
+  initialRawInputValue: string;
+  /** Initial input currency */
+  initialInputCurrency: Currency;
+  /** Currency to convert to */
+  initialOtherCurrency: Currency;
+};
+
+const toMoney = ({
+  value,
+  currency,
+}: {
+  value: string;
+  currency: Currency;
+}) => {
+  return new Money({
+    amount: value,
+    currency,
+    unit: getUnit(currency),
+  });
+};
+
+const getNumberOfDecimals = (value: string, decimalSeparator: string) => {
+  const decimalIndex = value.indexOf(decimalSeparator);
+  if (decimalIndex === -1) {
+    return 0;
+  }
+  return value.length - decimalIndex - 1;
+};
+
+const trimInputValue = (
+  value: string | undefined,
+  decimalSeparator: string,
+) => {
+  if (!value) {
+    return '0';
+  }
+
+  const decimalIndex = value.indexOf(decimalSeparator);
+  if (decimalIndex === -1) {
+    return value;
+  }
+
+  // Remove trailing zeros and decimal separator if it's the last character
+  return value
+    .replace(/0+$/, '')
+    .replace(new RegExp(`\\${decimalSeparator}$`), '');
+};
+
+export function useMoneyInput({
+  initialRawInputValue,
+  initialInputCurrency,
+  initialOtherCurrency,
+}: Props) {
+  const [state, setState] = useState<State>({
+    input: { value: initialRawInputValue, currency: initialInputCurrency },
+    converted: { value: undefined, currency: initialOtherCurrency },
+  });
+
+  const tickers: Ticker[] = [
+    `${state.input.currency}-${state.converted.currency}`,
+    `${state.converted.currency}-${state.input.currency}`,
+  ];
+
+  const { data: rates, error: exchangeRateError } = useExchangeRates(
```

Comment:
the key change was this one to load both main and inverse rate at one go so that swapping currencies doesn't trigger new rate fetch which would then trigger new conversion and that would result in loss of precision

the rest of the changes are mostly cleanup and renames to make it more clear what is going on. let me know if some stuff is still not clear and we can see if we can come up with better names